RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+#include <alpaka/alpaka.hpp>
+
+#include "FWCore/Utilities/interface/bit_cast.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/atomicMaxF.h"
+
+#include "DataFormats/ParticleFlowReco/interface/PFLayer.h"
+#include "RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.h"
+#include "RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterECLCC.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  using namespace cms::alpakatools;
+
+  using namespace reco::pfClustering;
+
+  static const int threadsPerBlockForClustering = std::is_same_v<Device, alpaka::DevCpu> ? 32 : 512;
+  static constexpr uint32_t blocksForExoticClusters = 4;
+
+  static constexpr uint32_t kHBHalf = 1296;
+  static constexpr uint32_t maxTopoInput = 2 * kHBHalf;
+
+  // Calculation of dR2 for Clustering
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE static float dR2(Position4 pos1, Position4 pos2) {
+    float mag1 = sqrtf(pos1.x * pos1.x + pos1.y * pos1.y + pos1.z * pos1.z);
+    float cosTheta1 = mag1 > 0.0 ? pos1.z / mag1 : 1.0;
+    float eta1 = 0.5 * logf((1.0 + cosTheta1) / (1.0 - cosTheta1));
+    float phi1 = atan2f(pos1.y, pos1.x);
+
+    float mag2 = sqrtf(pos2.x * pos2.x + pos2.y * pos2.y + pos2.z * pos2.z);
+    float cosTheta2 = mag2 > 0.0 ? pos2.z / mag2 : 1.0;
+    float eta2 = 0.5 * logf((1.0 + cosTheta2) / (1.0 - cosTheta2));
+    float phi2 = atan2f(pos2.y, pos2.x);
+
+    float deta = eta2 - eta1;
+    float dphi = std::abs(std::abs(phi2 - phi1) - M_PI) - M_PI;
```

Comment:
```suggestion
    constexpr const float fPI = M_PI; 
    float dphi = std::abs(std::abs(phi2 - phi1) - fPI) - fPI;
```
to avoid the intermediate conversions to double precision, which does not seem very useful since here it's doing only differences.